flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/reader/WatermarkExtractorRecordEmitter.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.reader;
+
+import org.apache.flink.api.common.eventtime.Watermark;
+import org.apache.flink.api.connector.source.SourceOutput;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Emitter which emits the record with event time and updates the split position.
+ *
+ * <p>The Emitter also emits watermarks at the beginning of every split, and sets the event
+ * timestamp based on the provided {@link IcebergWatermarkExtractor}.
+ */
+class WatermarkExtractorRecordEmitter<T> implements SerializableRecordEmitter<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(WatermarkExtractorRecordEmitter.class);
+  private final IcebergWatermarkExtractor timeExtractor;
+  private String lastSplit = null;
+  private long watermark;
+
+  WatermarkExtractorRecordEmitter(IcebergWatermarkExtractor timeExtractor) {
+    this.timeExtractor = timeExtractor;
+  }
+
+  @Override
+  public void emitRecord(
+      RecordAndPosition<T> element, SourceOutput<T> output, IcebergSourceSplit split) {
+    if (!split.splitId().equals(lastSplit)) {
+      long newWatermark = timeExtractor.extractWatermark(split);
+      if (newWatermark < watermark) {
+        LOG.warn(
+            "Watermark decreased. PreviousWM {}, currentWM {}, previousSplit {}, currentSplit {}.",
+            watermark,
+            newWatermark,
+            lastSplit,
+            split.splitId());
+      }
```

Comment:
Shouldn't Flink's watermark strategy determine if it's monotonic? Wondering if we would end up duplicating this logic in two places.

Reply:
If this is not monotonic, that would mean that we have a code issue.
We sort the splits by the watermark, so if a non-increasing watermark arrives, then it is a problem - at last by my understanding

Reply:
It seems like this approach assumes that all the splits are known during initialization. But what if a split is received late and has a watermark earlier than the current one? This could lead to emitting a warning, which can confuse the user, especially if the user's watermark strategy also emits the same warning.

Reply:
Let me put this on `info` level then.
TBH, I still think that having non-monotonically increasing watermark is an issue, but I am not sure how strict Flink in this regard. It is not too hard to use the max of the received watermarks, so Flink could do it, or the extractor could to it...